Can't save huge swagger api definition using ES storage

I'm using elasticsearch for my storage instead of the default RDBMS and everything worked fine. 

When I tried to save a new API definition for an API I created I noticed it doesn't save the definition. It does not show any error on the UI. I only notice the new definition did not reflect after refreshing the page. My new API definition is a bit large (160kb) which is greater than the 32kb size limit on elasticsearch keyword fields.

How can I change the API definition field to accept large JSON or YAML strings?

After reindexing, I keep getting StorageException errors all over the app. I can't do anything on the `System Administration` page and on several other pages as well.

`io.apiman.manager.api.core.exceptions.StorageException: java.lang.NullPointerException`

```io.apiman.manager.api.rest.contract.exceptions.SystemErrorException: io.apiman.manager.api.core.exceptions.StorageException: java.lang.NullPointerException
	at io.apiman.manager.api.rest.impl.GatewayResourceImpl.list(GatewayResourceImpl.java:117)
	at io.apiman.manager.api.rest.impl.GatewayResourceImpl$Proxy$_$$_WeldClientProxy.list(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:140)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:295)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:249)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:236)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:406)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:213)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:228)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:56)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:51)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:85)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at io.apiman.common.servlet.RootResourceFilter.doFilter(RootResourceFilter.java:59)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.apiman.manager.api.war.TransactionWatchdogFilter.doFilter(TransactionWatchdogFilter.java:57)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.apiman.manager.api.security.impl.DefaultSecurityContextFilter.doFilter(DefaultSecurityContextFilter.java:56)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.apiman.common.servlet.DisableCachingFilter.doFilter(DisableCachingFilter.java:52)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.apiman.common.servlet.ApimanCorsFilter.doFilter(ApimanCorsFilter.java:71)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.apiman.common.servlet.LocaleFilter.doFilter(LocaleFilter.java:61)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.keycloak.adapters.undertow.UndertowAuthenticatedActionsHandler.handleRequest(UndertowAuthenticatedActionsHandler.java:66)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:131)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.DisableCacheHandler.handleRequest(DisableCacheHandler.java:33)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AuthenticationConstraintHandler.handleRequest(AuthenticationConstraintHandler.java:53)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.servlet.handlers.security.ServletSecurityConstraintHandler.handleRequest(ServletSecurityConstraintHandler.java:59)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
	at org.keycloak.adapters.undertow.ServletPreAuthActionsHandler.handleRequest(ServletPreAuthActionsHandler.java:69)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:326)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:812)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.apiman.manager.api.core.exceptions.StorageException: java.lang.NullPointerException
	at io.apiman.manager.api.es.EsStorage.listEntities(EsStorage.java:2050)
	at io.apiman.manager.api.es.EsStorage.listGateways(EsStorage.java:1213)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.weld.bean.proxy.AbstractBeanInstance.invoke(AbstractBeanInstance.java:38)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:100)
	at io.apiman.manager.api.core.IStorageQuery$816713463$Proxy$_$$_WeldClientProxy.listGateways(Unknown Source)
	at io.apiman.manager.api.rest.impl.GatewayResourceImpl.list(GatewayResourceImpl.java:115)
	... 79 more
Caused by: java.lang.NullPointerException
	at io.searchbox.core.SearchResult.getHits(SearchResult.java:85)
	at io.searchbox.core.SearchResult.getHits(SearchResult.java:73)
	at io.searchbox.core.SearchResult.getHits(SearchResult.java:65)
	at io.searchbox.core.SearchResult.getHits(SearchResult.java:61)
	at io.apiman.manager.api.es.EsStorage.listEntities(EsStorage.java:2047)
	... 88 more```

--- a/common/es/src/main/resources-filtered/io/apiman/common/es/util/apiman_manager-settings.json
+++ b/common/es/src/main/resources-filtered/io/apiman/common/es/util/apiman_manager-settings.json
@@ -89,7 +89,7 @@
                 "entityVersion" : { "type" : "string", "index" : "not_analyzed" },
                 "who" : { "type" : "string", "index" : "not_analyzed" },
                 "what" : { "type" : "string", "index" : "not_analyzed" },
-                "data" : { "type" : "string", "index" : "not_analyzed" }
+                "data" : { "type" : "text" }
             }
         },
         "plan" : {

--- a/manager/api/es/src/main/java/io/apiman/manager/api/es/EsStorage.java
+++ b/manager/api/es/src/main/java/io/apiman/manager/api/es/EsStorage.java
@@ -84,6 +84,7 @@ import io.apiman.manager.api.es.util.XContentBuilder;
 import io.searchbox.action.Action;
 import io.searchbox.client.JestClient;
 import io.searchbox.client.JestResult;
+import io.searchbox.client.JestResultHandler;
 import io.searchbox.cluster.Health;
 import io.searchbox.core.Delete;
 import io.searchbox.core.DeleteByQuery;
@@ -96,7 +97,10 @@ import io.searchbox.core.SearchScroll;
 import io.searchbox.core.SearchScroll.Builder;
 import io.searchbox.core.search.sort.Sort;
 import io.searchbox.indices.CreateIndex;
+import io.searchbox.indices.DeleteIndex;
 import io.searchbox.indices.IndicesExists;
+import io.searchbox.indices.mapping.GetMapping;
+import io.searchbox.indices.reindex.Reindex;
 import io.searchbox.params.Parameters;
 
 import java.io.ByteArrayInputStream;
@@ -104,13 +108,7 @@ import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.URL;
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.List;
-import java.util.ListIterator;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
@@ -200,11 +198,136 @@ public class EsStorage implements IStorage, IStorageQuery {
                 JestResult result = esClient.execute(action);
                 if (!result.isSucceeded()) {
                     createIndex(getIndexName());
+                } else {
+                    if (isMigrationNeeded()){
+                        migrateIndexMapping();
+                    }
                 }
             }
         } catch (Exception e) {
             throw new RuntimeException(e);
         }
+    }
+
+    /**
+     * This method will check if a migration of the index is needed
+     * @return true or false
+     * @throws IOException
+     */
+    private boolean isMigrationNeeded() throws IOException {
+        // Check if migration is needed for too large swagger files - "keyword" to "text"
+        // See https://github.com/apiman/apiman/issues/736
+        JestResult result = esClient.execute(new GetMapping.Builder().addIndex(getIndexName()).build());
+        if (result.isSucceeded()) {
+            boolean migrationNeeded = result.getJsonObject()
+                    .getAsJsonObject(getIndexName())
+                    .getAsJsonObject("mappings")
+                    .getAsJsonObject("auditEntry")
+                    .getAsJsonObject("properties")
+                    .getAsJsonObject("data")
+                    .getAsJsonPrimitive("type")
+                    .getAsString()
+                    .equals("keyword");
+            if (migrationNeeded) {
+                logger.info("Migration of Elasticsearch index is needed.");
+                return true;
+            } else {
+                return false;
+            }
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * This method will migrate the indices and reindex them back to the original to change field mappings.
+     * We need this for https://github.com/apiman/apiman/issues/736
+     * @throws Exception
+     */
+    private void migrateIndexMapping() throws Exception {
+        logger.info("Migration of Elasticsearch index has started. This could take a moment.");
+        // create tmp index for reindex with new mapping
+        String tmpIndexName = getIndexName() + "_tmp";
+        createIndex(tmpIndexName);
+
+        // set source and dest index
+        HashMap<String, Object> source = new HashMap<>();
+        HashMap<String, Object> dest = new HashMap<>();
+        source.put("index", getIndexName());
+        dest.put("index", tmpIndexName);
+
+
+        // reindex old index to tmp index
+        Action<JestResult> reindexToTmp = new Reindex.Builder(source, dest).waitForActiveShards(1).waitForCompletion(true).refresh(true).build();
+        esClient.executeAsync(reindexToTmp, new JestResultHandler<JestResult>() {
+            @Override
+            public void completed(JestResult jestResult) {
+                logger.info("Reindex to " + tmpIndexName);
+                logger.info("Result: " + jestResult.getJsonString());
+
+                // delete old index
+                Action<JestResult> deleteIndex = new DeleteIndex.Builder(getIndexName()).build();
+                esClient.executeAsync(deleteIndex, new JestResultHandler<JestResult>() {
+                    @Override
+                    public void completed(JestResult jestResult) {
+                        logger.info("Deleted: " + getIndexName());
+                        logger.info("Result: " + jestResult.getJsonString());
+
+                        // create a new index with old name and new mapping
+                        try {
+                            createIndex(getIndexName());
+                        } catch (Exception e) {
+                            throw new RuntimeException(e);
+                        }
+
+                        // redindex tmp index back to "old" index
+                        source.put("index", tmpIndexName);
+                        dest.put("index", getIndexName());
+
+                        Action<JestResult> reindexTmpToNew = new Reindex.Builder(source, dest).waitForActiveShards(1).waitForCompletion(true).refresh(true).build();
+                        esClient.executeAsync(reindexTmpToNew, new JestResultHandler<JestResult>() {
+                            @Override
+                            public void completed(JestResult jestResult) {
+                                logger.info("Reindex to " + getIndexName());
+                                logger.info("Result: " + jestResult.getJsonString());
+
+                                // delete tmp index
+                                Action<JestResult> deleteTmpIndex = new DeleteIndex.Builder(tmpIndexName).build();
+                                esClient.executeAsync(deleteTmpIndex, new JestResultHandler<JestResult>() {
+                                    @Override
+                                    public void completed(JestResult jestResult) {
+                                        logger.info("Deleted: " + tmpIndexName);
+                                        logger.info("Result: " + jestResult.getJsonString());
+                                        logger.info("Migration succeeded");
+                                    }
+
+                                    @Override
+                                    public void failed(Exception e) {
+                                        throw new RuntimeException(e);
+                                    }
+                                });
+                            }
+
+                            @Override
+                            public void failed(Exception e) {
+                                throw new RuntimeException(e);
+                            }
+                        });
+                    }
+
+                    @Override
+                    public void failed(Exception e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+            }
+
+            @Override
+            public void failed(Exception e) {
+                throw new RuntimeException(e);
+            }
+        });
+
     }
 
     /**
@@ -217,6 +340,8 @@ public class EsStorage implements IStorage, IStorageQuery {
         JestResult response = esClient.execute(new CreateIndex.Builder(indexName).settings(source).build());
         if (!response.isSucceeded()) {
             throw new StorageException("Failed to create index " + indexName + ": " + response.getErrorMessage()); //$NON-NLS-1$ //$NON-NLS-2$
+        } else {
+            logger.info("Created index:" + indexName);
         }
     }
 

--- a/manager/api/es/src/main/resources/io/apiman/manager/api/es/index-settings.json
+++ b/manager/api/es/src/main/resources/io/apiman/manager/api/es/index-settings.json
@@ -89,7 +89,7 @@
                 "entityVersion" : { "type" : "string", "index" : "not_analyzed" },
                 "who" : { "type" : "string", "index" : "not_analyzed" },
                 "what" : { "type" : "string", "index" : "not_analyzed" },
-                "data" : { "type" : "string", "index" : "not_analyzed" }
+                "data" : { "type" : "text" }
             }
         },
         "plan" : {

